pkg/controllers/cluster/actions/upgrade_version.go

```diff
@@ -3,54 +3,966 @@ package actions
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	a.Cluster = patched
+
+	return err
+}
+
+func (a *ClusterVersionUpgrade) onStateTransition(ctx context.Context, currentState, nextState fsm.StateType, event fsm.EventType) error {
+	a.logger.Info(ctx, "Upgrade state transition", "event", event, "from", currentState, "to", nextState)
+
+	err := a.upgradeClusterStatus(ctx, func(status *scyllav1alpha1.ClusterStatus) {
+		status.Upgrade.State = string(nextState)
+	})
+	if err != nil {
+		return errors.Wrap(err, "patch cluster status")
+	}
+
+	return nil
+}
+
+func (a *ClusterVersionUpgrade) finishUpgrade(ctx context.Context) (fsm.EventType, error) {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		cluster := &scyllav1alpha1.ScyllaCluster{}
+		if err := a.cc.Get(ctx, naming.NamespacedName(a.Cluster.Name, a.Cluster.Namespace), cluster); err != nil {
+			return err
+		}
+		cluster.Status.Upgrade = nil
+		if err := a.cc.Status().Update(ctx, cluster); err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return ActionFailure, errors.Wrap(err, "update cluster status")
+	}
+
+	a.recorder.Event(a.Cluster, corev1.EventTypeNormal, naming.SuccessSynced, fmt.Sprintf("Cluster upgraded to version %s", a.Cluster.Spec.Version))
+
+	// Return NoOp which will stop state machine.
+	return fsm.NoOp, nil
+}
+
+func (a *ClusterVersionUpgrade) setRollingUpgradeStrategy(ctx context.Context, rack *scyllav1alpha1.RackSpec) error {
+	sts := &appsv1.StatefulSet{}
+	err := a.cc.Get(ctx, naming.NamespacedName(naming.StatefulSetNameForRack(*rack, a.Cluster), a.Cluster.Namespace), sts)
+	if err != nil {
+		return errors.Wrap(err, "get rack sts")
+	}
+
+	upgradeSts := sts.DeepCopy()
+	upgradeSts.Spec.UpdateStrategy = appsv1.StatefulSetUpdateStrategy{
+		Type: appsv1.RollingUpdateStatefulSetStrategyType,
+	}
+
+	if err := util.PatchStatefulSet(ctx, sts, upgradeSts, a.kubeClient); err != nil {
+		return errors.Wrap(err, "update rack sts")
+	}
+	return nil
+}
+
+func (a *ClusterVersionUpgrade) restoreUpgradeStrategy(ctx context.Context) (fsm.EventType, error) {
+	rack, err := a.getCurrentRack(ctx)
+	if err != nil {
+		return ActionFailure, errors.Wrap(err, "get rack")
+	}
+
+	if rack != nil {
+		if err := a.setRollingUpgradeStrategy(ctx, rack); err != nil {
+			return ActionFailure, errors.Wrap(err, "set upgrade strategy")
+		}
+	} else {
+		// When all nodes upgraded, validate if every has proper upgrade strategy
+		for _, r := range a.Cluster.Spec.Datacenter.Racks {
+			if err := a.setRollingUpgradeStrategy(ctx, &r); err != nil {
+				return ActionFailure, errors.Wrap(err, "set upgrade strategy")
+			}
+		}
+	}
+
+	return ActionSuccess, nil
+}
+
+func (a *ClusterVersionUpgrade) fsm() *fsm.StateMachine {
+	state := BeginUpgrade
+	if a.Cluster.Status.Upgrade != nil {
+		state = fsm.StateType(a.Cluster.Status.Upgrade.State)
+	}
+
+	return &fsm.StateMachine{
```

Comment:
It looks to me that the FSM is a plain list. Maybe it would be easier to drop the FSM and just proceed with the list of steps.

I cannot see support for ActionFailure, there is not cleanup path?

Reply:
> I cannot see support for ActionFailure, there is not cleanup path?

`ActionFailure fsm.Event = fsm.NoOp `
failed action will be retried on next reconcilation iteration. All actions except beginUpgrade should be idempotent, so upgrade is always retried, and there's no actual failure path. 


> It looks to me that the FSM is a plain list. Maybe it would be easier to drop the FSM and just proceed with the list of steps.

FSM isn't flat, there two loops:
FindNextNode->EnableMaintenanceMode->[...]->ClearDataBackup->FindNextNode
and 
FindNextRack->[...]->RestoreUpgradeStrategy->FindNextRack